tools/lintapidiff.ml

```diff
@@ -0,0 +1,268 @@
+(* Detects newly added symbols that are missing "@since" annotations,
+   or removed symbols that didn't have "@deprecated" annotation before.
+
+   Handles: values, exceptions.
+   Ignores: variants, record fields, classes, module aliasing or includes, ...
+   Out of scope: changes in arity, parameters, ...
+
+   Missing attributes on undocumented identifiers in undocumented modules
+   are not reported.
+
+   Use lintapidiff.sh to run
+*)
+open Location
+open Parsetree
+
+module IdMap = Depend.StringMap
+type version = string
+
+module Doc = struct
+  type t = {
+    since: version option;
+    deprecated: bool;
+    loc: Location.t;
+    has_doc_parent: bool;
+    has_doc: bool;
+  }
+
+  let empty = {since = None; deprecated=false; loc=Location.none;
+               has_doc_parent=false;has_doc=false}
+
+  let since = Str.regexp "\\(.\\|\n\\)*@since +\\([^ ]+\\).*"
+
+  let find_attr lst attrs =
+    try Some (List.find (fun (loc, _) -> List.mem loc.txt lst) attrs)
+    with Not_found -> None
+
+  let get_doc lst attrs = match find_attr lst attrs with
+    | Some (_, PStr [{pstr_desc=Pstr_eval(
+        {pexp_desc=Pexp_constant(Pconst_string (doc, _));_}, _);_}])
+      when doc <> "/*" && doc <> "" -> Some doc
+    | _ -> None
+
+  let is_deprecated attrs =
+    find_attr ["ocaml.deprecated"; "deprecated"] attrs <> None ||
+    match get_doc ["ocaml.text"] attrs with (* for toplevel module annotation *)
+    | None -> false
+    | Some text ->
+        try Misc.search_substring "@deprecated" text 0 >= 0;
+        with Not_found -> false
+
+  let get parent_info loc attrs =
+    let doc = get_doc ["ocaml.doc"; "ocaml.text"] attrs in
+    {
+      since = (match doc with
+          | Some doc ->
+              if Str.string_match since doc 0 then
+                Some (Str.matched_group 2 doc |> String.trim)
+              else parent_info.since
+          | None -> parent_info.since);
+      deprecated = parent_info.deprecated || is_deprecated attrs;
+      loc;
+      has_doc_parent = parent_info.has_doc;
+      has_doc = doc <> None
```

Comment:
Making `has_doc_parent` fully propagated (i.e. `parent_info.has_doc_parent || parent_info.has_doc`) produces a lot of noise for MoreLabels, although I could easily just exclude morelabels from checking, and enable full propagation.